File: base/cpp/test-set.cpp
#include <iostream>
#include "aps-impl.h"
#include "basic.h"

using namespace std;

typedef C_SET<C_INTEGER> C_IntegerSet;
typedef C_IntegerSet::T_Result T_IntegerSet;

static bool verbose = false;
static int failed = 0;

template <class T>
static void assert_equals(const string &test, T expected, T actual)
{
  bool passed = actual == expected;
  if (verbose || !passed) cout << test << " = " << actual << endl;
  if (passed) return;
  ++failed;
  cout << "  failed!  (Expected " << expected << ")" << endl;
}

int main(int argc, char **argv)
{
  if (argc > 1) verbose = true;
  T_IntegerSet is, is2, is3;
  C_IntegerSet* t_IntegerSet = new C_IntegerSet(get_Integer());
  COLL<C_IntegerSet,C_Integer> iscoll(t_IntegerSet,get_Integer());
  C_IntegerSet& t = *t_IntegerSet;
  is = t.v_append(t.v_append(t.v_single(1),t.v_single(2)),t.v_single(3));
  is2 = t.v_append(t.v_append(t.v_single(3),t.v_single(0)),t.v_single(2));
  is3 = t.v_append(t.v_append(t.v_single(3),t.v_single(1)),t.v_single(2));

  assert_equals("s1",string("{1,2,3}"),iscoll.to_string(is));
  assert_equals("s2",string("{3,0,2}"),iscoll.to_string(is2));
  assert_equals("s3",string("{3,1,2}"),iscoll.to_string(is3));

  assert_equals("union(s1,s2)",string("{1,3,0,2}"),
		iscoll.to_string(t.v_union(is,is2)));
  assert_equals("intersect(s1,s2)",string("{2,3}"),
		iscoll.to_string(t.v_intersect(is,is2)));
  assert_equals("difference(s1,s2)",string("{1}"),
		iscoll.to_string(t.v_difference(is,is2)));

  assert_equals("equal(s1,s2)",false,t.v_equal(is,is2));
  assert_equals("equal(s1,s3)",true,t.v_equal(is,is3));
  assert_equals("equal(s2,s3)",false,t.v_equal(is2,is3));
  
  return failed;
}

File: base/scala/test-set.scala
import basic_implicit._;

object TestSet extends App {
  val t_Set = new M_SET[T_Integer]("SET", t_Integer);

  val l1 = Set(1,2,3,2,1,4,5,4,3,2,2,2);
  val l2 = Set(0,1,2,3,4,5,2);
  val l1_clone = l1 ++ Set();

  println("union = "+ t_Set.v_union(l1,l2));
  println("intersect = " + t_Set.v_intersect(l1,l2));
  println("difference = " + t_Set.v_difference(l1,l2));
  println("combine = "+ t_Set.v_combine(l1,l2));
  println("equal(l1,l1) = " + t_Set.v_equal(l1,l1));
  println("equal(l1,l1_clone) = " + t_Set.v_equal(l1,l1_clone));
  println("equal(l1,l2) = " + t_Set.v_equal(l1,l2));
}
